providers/common/include/prov/der.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/bn.h>
+
+/* Well known primitive tags */
+
+#define DER_P_EOC                       0 /* BER End Of Contents tag */
+#define DER_P_BOOLEAN                   1
+#define DER_P_INTEGER                   2
+#define DER_P_BIT_STRING                3
+#define DER_P_OCTET_STRING              4
+#define DER_P_NULL                      5
+#define DER_P_OBJECT                    6
+#define DER_P_OBJECT_DESCRIPTOR         7
+#define DER_P_EXTERNAL                  8
+#define DER_P_REAL                      9
+#define DER_P_ENUMERATED               10
+#define DER_P_UTF8STRING               12
+#define DER_P_SEQUENCE                 16
+#define DER_P_SET                      17
+#define DER_P_NUMERICSTRING            18
+#define DER_P_PRINTABLESTRING          19
+#define DER_P_T61STRING                20
+#define DER_P_VIDEOTEXSTRING           21
+#define DER_P_IA5STRING                22
+#define DER_P_UTCTIME                  23
+#define DER_P_GENERALIZEDTIME          24
+#define DER_P_GRAPHICSTRING            25
+#define DER_P_ISO64STRING              26
+#define DER_P_GENERALSTRING            27
+#define DER_P_UNIVERSALSTRING          28
+#define DER_P_BMPSTRING                30
+
+#define DER_F_PRIMITIVE              0x00
+#define DER_F_CONSTRUCTED            0x20
+
+#define DER_C_UNIVERSAL              0x00
+#define DER_C_APPLICATION            0x40
+#define DER_C_CONTEXT                0x80
+#define DER_C_PRIVATE                0xC0
+
+/* Helper macros */
+#define CHECKEDret(l, r, f)                     \
```

Comment:
Possibly resolve these inline too.

I didn't think flow of control constructs we liked in multi-line macro expansions (`return (r)`).

Reply:
What do you mean with "resolve these lines"?

Reply:
Hmmm, I could actually go without this macro by passing in a reference to the counter along with the pointer reference and the starting point, and then just returning a boolean, allowing for out typical cascading condition.  Yeahok, that's more OpenSSL style.

Reply:
So what I mean is that code that currently looks like this:

``` C
size_t DER_w_something(unsigned char **pp, const unsigned char *start,
                       SOMETHING *s)
{
    size_t a = 0;

    CHECKED(a, DER_w_ulong(pp, start, v));
    CHECKED(a, DER_w_precompiled(pp, start, OID, sizeof(OID));
    return a;
}
```

would look like this instead:

``` C
int DER_w_something(unsigned char **pp, size_t *cnt, const unsigned char *start,
                    SOMETHING *s)
{
    if (!DER_w_ulong(pp, cnt, start, v)
        || !DER_w_precompiled(pp, start, OID, sizeof(OID))
        return 0;
    return a;
}
```

There's a bit of elegance lost there, but like I realised, it's more "OpenSSL"

Reply:
There's a complicating factor in the "OpenSSL" style API...  Any time a function writes a constructed structure, such as a sequence, the author must remember to update the referred to counter correctly.  For example, this:

``` C
size_t DER_w_algorithmIdentifier_DSA(unsigned char **pp,
                                     const unsigned char *start,
                                     DSA *dsa)
{
    size_t a = 0;

    /* No parameters */
    CHECKED(a, DER_w_precompiled(pp, start, der_oid_id_dsa,
                                 sizeof(der_oid_id_dsa)));
    CHECKED(a, DER_w_sequence(pp, start, a));
    return a;
}
```

becomes this:

``` C
int DER_w_algorithmIdentifier_DSA(unsigned char **pp, size_t *cnt,
                                  const unsigned char *start,
                                  DSA *dsa)
{
    size_t a = 0;

    /* No parameters (yet?) */
    if (!DER_w_precompiled(pp, &a, start, der_oid_id_dsa,
                           sizeof(der_oid_id_dsa))
        || !DER_w_sequence(pp, &a, start, a))
        return 0;
    *cnt += a;
    return 1;
}
```

....

You *could* also write it like this, but then it becomes less obvious and opens up for more programming mistakes:

``` C
int DER_w_algorithmIdentifier_DSA(unsigned char **pp, size_t *cnt,
                                  const unsigned char *start,
                                  DSA *dsa)
{
    size_t a = 0;

    /* No parameters (yet?) */
    if (!DER_w_precompiled(pp, &a, start, der_oid_id_dsa,
                           sizeof(der_oid_id_dsa))
        || !DER_w_sequence(pp, cnt, start, a))        /* <-- see what happens here? */
        return 0;
    return 1;
}
```

Reply:
....

Moaning aside, I think I can get used to the changed API.

Reply:
I'd prefer the CHECKED approach as more readable.

Reply:
I would too, for that reason.  However, there's readable and there's readable; CHECKED relies on control flow "magic", which means you have to know exactly what CHECKED does before you use it, and there are differing views on what's actually readable.

So far, the more aesthetic viewpoint (in which I count retaining CHECKED) hasn't been the winning one (we had exactly this kind of discussion around OpenSSL assertions, as I recall it)

Reply:
API changed.